pkg/receive/handler.go

```diff
@@ -747,34 +758,65 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, wreqs map[e
 	// This is needed if context is canceled or if we reached success of fail quorum faster.
 	defer func() {
 		go func() {
-			for err := range ec {
+			for errReplicaFn := range ec {
+				_, _, err := errReplicaFn()
 				if err != nil {
 					level.Debug(tLogger).Log("msg", "request failed, but not needed to achieve quorum", "err", err)
 				}
 			}
 		}()
 	}()
 
-	var success int
+	var (
+		// perReplicaSuccess is used to determine if enough requests succeeded for given replica number.
+		preReplicaSuccess = make([]int, h.options.ReplicationFactor)
```

Comment:
Shouldn't we be tracking per-shard success rate?

Reply:
What do you mean with "shard", @fpetkovski? Is is the same as the "replica group request" mentioned in the description of this pull request?

Reply:
I think we might be talking about the same thing, but I still think this variable is very confusing. Requests get split into independent partitions (shards), and then each shard gets replicated. So I would expect us to track how many successful replications each shard has, e.g `perShardSuccessfulReplicas`.

Reply:
I'm not sure I follow fully as well, but throwing in yet another term is even more confusing for me. Are we talking about the initial batching of requests in https://github.com/thanos-io/thanos/blob/main/pkg/receive/handler.go#L555 when you say 'shards'? Since most of the docs refer to these as 'batches', I'd either stick to this or refer to this as 'shards' everywhere (I don't have strong preference but I associate shards more with actual databases rather than requests).

Other than that you're right, but I named it 'per replica', since each index in the slice tracks success of requests going to given replica (for a _single_ shard / batch), but if it's not obvious enough I can prefix it as you suggested.

Reply:
Calling it a batch makes sense as well, I agree that sharding is not mentioned anywhere else.

Reply:
If I'm not mistaken, in most of the places in this part of the code we could literally replace the world "replica" with "replicationIndex". This confuses me a lot.

Reply:
Agreed @douglascamata, I think this where lot of the confusion originates, as @fpetkovski also mentions, this creates impression that replica = receiver node, but it's more of a logical replica (since parts of that replica will end up on different nodes). I'd be happy to rename this, I'll also try to improve the other variables and get rid of the "group request" naming.

Reply:
Amaaaaaazing, @matej-g!